src/main/java/org/apache/sysds/runtime/instructions/fed/CovarianceFEDInstruction.java

```diff
@@ -0,0 +1,253 @@
+		String opcode, String istr) {
+		super(FEDInstruction.FEDType.AggregateBinary, op, in1, in2, in3, out, opcode, istr);
+	}
+
+
+	public static CovarianceFEDInstruction parseInstruction(String str) {
+		CPOperand in1 = new CPOperand("", Types.ValueType.UNKNOWN, Types.DataType.UNKNOWN);
+		CPOperand in2 = new CPOperand("", Types.ValueType.UNKNOWN, Types.DataType.UNKNOWN);
+		CPOperand in3 = null;
+		CPOperand out = new CPOperand("", Types.ValueType.UNKNOWN, Types.DataType.UNKNOWN);
+
+		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
+		String opcode = parts[0];
+
+		if( !opcode.equalsIgnoreCase("cov") ) {
+			throw new DMLRuntimeException("CovarianceCPInstruction.parseInstruction():: Unknown opcode " + opcode);
+		}
+
+		COVOperator cov = new COVOperator(COV.getCOMFnObject());
+		if ( parts.length == 4 ) {
+			parseBinaryInstruction(str, in1, in2, out);
+			return new CovarianceFEDInstruction(cov, in1, in2, out, opcode, str);
+		} else if ( parts.length == 5 ) {
+			in3 = new CPOperand("", Types.ValueType.UNKNOWN, Types.DataType.UNKNOWN);
+			parseBinaryInstruction(str, in1, in2, in3, out);
+			return new CovarianceFEDInstruction(cov, in1, in2, in3, out, opcode, str);
+		}
+		else {
+			throw new DMLRuntimeException("Invalid number of arguments in Instruction: " + str);
+		}
+	}
+
+	@Override
+	public void processInstruction(ExecutionContext ec) {
+		MatrixObject mo1 = ec.getMatrixObject(input1);
+		MatrixObject mo2 = ec.getMatrixObject(input2);
+		COVOperator cop = ((COVOperator)_optr);
+
+		if(mo1.isFederated() && mo2.isFederated() && !mo1.getFedMapping().isAligned(mo2.getFedMapping(), false))
+			throw new DMLRuntimeException("Not supported matrix-matrix binary operation: covariance.");
+
+		FederatedRequest fr2 = null;
+		if(mo1.isFederated() && mo2.isFederated() && mo1.getFedMapping().isAligned(mo2.getFedMapping(), false)) {
+			fr2 = FederationUtils.callInstruction(instString, output, new CPOperand[]{input1, input2},
+				new long[]{mo1.getFedMapping().getID(), mo2.getFedMapping().getID()});
+			Future<FederatedResponse>[] tmp = mo1.getFedMapping().execute(getTID(), true, fr2);
+
+			//means
+			Future<FederatedResponse>[] meanTmp1 = processMean(mo1, 0);
+			Future<FederatedResponse>[] meanTmp2 = processMean(mo2, 1);
+
+			double res = aggCov(tmp, meanTmp1, meanTmp2, mo1.getFedMapping());
+			ec.setVariable(output.getName(), new DoubleObject(res));
+		}
+		else {
+			MatrixBlock mb;
+			MatrixObject mo;
+			if(!mo1.isFederated() && mo2.isFederated()) {
+				mo = mo2;
+				mb = ec.getMatrixInput(input1.getName());
+			}
+			else {
+				mo = mo1;
+				mb = ec.getMatrixInput(input2.getName());
+			}
+
+			FederationMap fedMapping = mo.getFedMapping();
+			List<CM_COV_Object> globalCmobj = new ArrayList<>();
+			long varID = FederationUtils.getNextFedDataID();
+
+			fedMapping.mapParallel(varID, (range, data) -> {
```

Comment:
map parallel, using a function on the right side of the lambda would also be cool here.